src/mgxs_data.F90

```diff
@@ -0,0 +1,705 @@
+                               error_text)
+
+          ! Keep track of what listing is associated with this nuclide
+          nuclides_MG(i_nuclide) % obj % listing = i_listing
+
+          ! Handle any errors
+          if (error_code /= 0) then
+            call fatal_error(trim(error_text))
+          end if
+
+          ! Add name and alias to dictionary
+          call already_read % add(name)
+          call already_read % add(alias)
+        end if
+      end do NUCLIDE_LOOP
+    end do MATERIAL_LOOP
+
+    ! Avoid some valgrind leak errors
+    call already_read % clear()
+
+    ! Loop around material
+    MATERIAL_LOOP3: do i = 1, n_materials
+
+      ! Get material
+      mat => materials(i)
+
+      ! Loop around nuclides in material
+      NUCLIDE_LOOP2: do j = 1, mat % n_nuclides
+        ! Is this fissionable?
+        if (nuclides_MG(mat % nuclide(j)) % obj % fissionable) then
+          mat % fissionable = .true.
+        end if
+        if (mat % fissionable) then
+          exit NUCLIDE_LOOP2
+        end if
+
+      end do NUCLIDE_LOOP2
+    end do MATERIAL_LOOP3
+
+  end subroutine read_mgxs
+
+!===============================================================================
+! SAME_NUCLIDE_LIST creates a linked list for each nuclide containing the
+! indices in the nuclides array of all other instances of that nuclide.  For
+! example, the same nuclide may exist at multiple temperatures resulting
+! in multiple entries in the nuclides array for a single zaid number.
+!===============================================================================
+
+  subroutine same_NuclideMG_list()
+
+    integer :: i ! index in nuclides array
+    integer :: j ! index in nuclides array
+
+    do i = 1, n_nuclides_total
+      do j = 1, n_nuclides_total
+        if (nuclides_MG(i) % obj % zaid == nuclides_MG(j) % obj % zaid) then
+          call nuclides_MG(i) % obj % nuc_list % push_back(j)
+        end if
+      end do
+    end do
+
+  end subroutine same_NuclideMG_list
+
+!===============================================================================
+! NUCLIDE_*_INIT reads in the data from the XML file, as already accessed
+!===============================================================================
+
+    subroutine NuclideMG_init(this, node_xsdata, groups, get_kfiss, get_fiss, &
+                                  error_code, error_text)
+      class(NuclideMG), intent(inout) :: this        ! Working Object
+      type(Node), pointer, intent(in)  :: node_xsdata ! Data from data.xml
+      integer, intent(in)              :: groups      ! Number of Energy groups
+      logical, intent(in)              :: get_kfiss   ! Need Kappa-Fission?
+      logical, intent(in)              :: get_fiss ! Should we get fiss data?
+      integer, intent(inout)           :: error_code  ! Code signifying error
+      character(MAX_LINE_LEN), intent(inout) :: error_text ! Error message to print
```

Comment:
Error code/text can be `intent(out)` unless I'm mistaken... And was there a reason you wanted to return a code specifically as opposed to calling fatal_error?